bootloader/piboot.go

```diff
@@ -0,0 +1,421 @@
+			// config.txt that will allow booting.
+			trybootPath := filepath.Join(ubuntuSeedDir, "tryboot.txt")
+			if err := os.Remove(trybootPath); err != nil {
+				logger.Noticef("cannot remove %s: %v", trybootPath, err)
+			}
+		}
+	}
+
+	if dirtyEnv {
+		if err := env.Save(); err != nil {
+			return err
+		}
+	}
+
+	if reconfigBootloader {
+		if err := p.loadAndApplyConfig(env); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) SetBootVarsFromInitramfs(values map[string]string) error {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return err
+	}
+
+	dirtyEnv := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirtyEnv = true
+	}
+
+	if dirtyEnv {
+		if err := env.Save(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) loadAndApplyConfig(env *ubootenv.Env) error {
+	var prefix, cfgDir, dstDir string
+
+	cfgFile := "config.txt"
+	if env.Get("snapd_recovery_mode") == "run" {
+		kernelSnap := env.Get("snap_kernel")
+		kernStat := env.Get("kernel_status")
+		if kernStat == "try" {
+			// snap_try_kernel will be set when installing a new kernel
+			kernelSnap = env.Get("snap_try_kernel")
+			cfgFile = "tryboot.txt"
+		}
+		prefix = filepath.Join(pibootPartFolder, kernelSnap)
+		cfgDir = ubuntuSeedDir
+		dstDir = filepath.Join(ubuntuSeedDir, prefix)
+	} else {
+		// install/recovery modes, use recovery kernel
+		prefix = filepath.Join("/systems", env.Get("snapd_recovery_system"),
+			"kernel")
+		cfgDir = p.rootdir
+		dstDir = filepath.Join(p.rootdir, prefix)
+	}
+
+	logger.Debugf("configure piboot %s with prefix %q, cfgDir %q, dstDir %q",
+		cfgFile, prefix, cfgDir, dstDir)
+
+	if err := os.MkdirAll(dstDir, 0755); err != nil {
+		return err
+	}
+	return p.applyConfig(env, cfgFile, prefix, cfgDir, dstDir)
+}
+
+// Writes os_prefix in RPi config.txt or tryboot.txt
+func (p *piboot) writeRPiCfgWithOsPrefix(prefix, inFile, outFile string) error {
+	buf, err := ioutil.ReadFile(inFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+
+	replaced := false
+	newOsPrefix := "os_prefix=" + prefix + "/"
+	for i, line := range lines {
+		if strings.HasPrefix(line, "os_prefix=") {
+			if replaced {
+				logger.Noticef("unexpected extra os_prefix line: %q", line)
+				lines[i] = "# " + lines[i]
+				continue
+			}
+			lines[i] = newOsPrefix
+			replaced = true
+		}
+	}
+	if !replaced {
+		lines = append(lines, newOsPrefix)
+		lines = append(lines, "")
+	}
+
+	output := strings.Join(lines, "\n")
+	return osutil.AtomicWriteFile(outFile, []byte(output), 0644, 0)
+}
+
+func (p *piboot) writeCmdline(env *ubootenv.Env, defaultsFile, outFile string) error {
+	buf, err := ioutil.ReadFile(defaultsFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+	cmdline := lines[0]
```

Comment:
does cmdline.txt follow enough of the format of the actual kernel command line that we could use `osutil.KernelCommandLineSplit` ?

Reply:
perhaps that function suffers from the problem that we don't have a great way of serializing a map of kernel command line parameters and values/keys into a string to be written out again

Reply:
do we accumulate options though when you only ever append and don't unset previous values earlier in the string? i.e. I'm wondering if we will have the issue where we get:

```
foo=bar quiet splash kernel_status=trying kernel_status=trying kernel_status=trying
```
etc. 

I don't think we have the same problem for snapd_recovery_system but I'm not really sure

Reply:
note that AFAICT the reference file is always coming from the real runtime directory, not from the gadget (unless I am missing something), if we did use the gadget as the reference this wouldn't happen since the gadget version of cmdline.txt would be read-only and shouldn't (maybe we should verify) have `kernel_status` set in it, but now if we are adding a kernel_status=trying to the new cmdline.txt for a tried kernel dir on ubuntu-seed, what removes that string from the file?

Reply:
The `cmdline.txt` that is being used here is from the root of the seed partition (it is copied there when the gadget is installed), and the one that is written is in the directory specified by `os_prefix`, for instance `systems/20220302/kernel/cmdline.txt` when preparing the image, or in the corresponding directory when a new kernel is installed. So, the file in the root is not being rewritten, it is only read. And when status is changed, the destination `<folder>/cmdline.txt` is fully rewritten, including the previous kernel_status.

Reply:
thanks for explaining that, this all makes sense then and this particualr code lgtm then